### PR TITLE
security: use rtrim, not unsafe /X+$/

### PR DESCRIPTION
**Pull request by davisjam**
Problem:
`replace(/X+$/, '')` is vulnerable to REDOS

Solution:
Replace these instances with a custom rtrim.

### CODE REVIEW

**Comment by davisjam** [comment]
This is an excerpt from #1226. It affects "clean-up" regexes, not "parsing" regexes.

File: lib/marked.js [hunk 1]
```diff
@@ -1355,6 +1355,38 @@ function splitCells(tableRow, count) {
...
 
+// Return str with all trailing {c | all but c} removed
+// allButC: Default false
+function rtrim(str, c, allButC) {
+  if (typeof allButC === 'undefined') {
+    allButC = false;
+  } else {
+    allButC = true;
+  }
+  var mustMatchC = !allButC;
```
[the remark this answers is not in the document]
Comment (davisjam) [reply lib/marked.js]: I'll tweak the naming.

**Comment by davisjam** [comment]
(I am on vacation until end of May so won't be acting on this until then @joshbruce @styfle @UziTech).

File: lib/marked.js [hunk 2]
```diff
@@ -1355,6 +1355,38 @@ function splitCells(tableRow, count) {
   return cells;
 }
 
+// Remove trailing 'c's. Equivalent to str.replace(/c*$/, '').
+// /c*$/ is vulnerable to REDOS.
+// invert: Remove suffix of non-c chars instead. Default false.
+function rtrim(str, c, invert) {
+  if (typeof invert === 'undefined' || !invert) {
```
Comment (styfle) [inline lib/marked.js]: Is this `if` statement necessary?

Comment (davisjam) [reply lib/marked.js]: It appears not. Too many years of C have addled my brain.

**Comment by davisjam** [comment]
@styfle 0610f9f removes the `if` check you flagged.

**Pull request merged**